core/kernel/api/class.ModelFactory.php

```diff
@@ -16,112 +17,146 @@
+                    $returnValue |= $this->addStatement($modelId, $subjectUri, $prop, $v['value'], isset($v['lang']) ? $v['lang'] : null);
                 }
             }
         }
-        
-        return true;
+
+        return $returnValue;
     }
-    
+
     /**
      * Adds a statement to the ontology if it does not exist yet
-     * 
+     *
      * @author "Joel Bout, <joel@taotesting.com>"
-     * @param int $modelId
+     *
+     * @param int    $modelId
      * @param string $subject
      * @param string $predicate
      * @param string $object
      * @param string $lang
+     * @param string $author
+     *
+     * @return bool Was a row added?
      */
-    private function addStatement($modelId, $subject, $predicate, $object, $lang = null) {
-        $result = core_kernel_classes_DbWrapper::singleton()->query(
+    public function addStatement($modelId, $subject, $predicate, $object, $lang = null, $author = self::DEFAULT_AUTHOR)
+    {
+        // TODO: refactor this to use a triple store abstraction.
+        $result = $this->dbWrapper->query(
             'SELECT count(*) FROM statements WHERE modelid = ? AND subject = ? AND predicate = ? AND object = ? AND l_language = ?',
-            array($modelId, $subject, $predicate, $object, (is_null($lang)) ? '' : $lang)
+            [$modelId, $subject, $predicate, $object, (is_null($lang)) ? '' : $lang]
         );
-        
-        if (intval($result->fetchColumn()) === 0) {
-            $dbWrapper = core_kernel_classes_DbWrapper::singleton();
-            $date = $dbWrapper->getPlatForm()->getNowExpression();
-
-            $dbWrapper->insert(
-                'statements',
-                array(
-                    'modelid' =>  $modelId,
-                    'subject' =>$subject,
-                    'predicate'=> $predicate,
-                    'object' => $object,
-                    'l_language' => is_null($lang) ? '' : $lang,
-                    'author' => 'http://www.tao.lu/Ontologies/TAO.rdf#installator',
-                    'epoch' => $date
-                )
-            );
+
+        if (intval($result->fetchColumn()) > 0) {
+            return false;
         }
+
+        $date = $this->dbWrapper->getPlatForm()->getNowExpression();
+
+        return (bool) $this->dbWrapper->insert(
+            'statements',
+            [
+                'id' => $this->getUniquePrimaryKey(),
+                'modelid' => $modelId,
+                'subject' => $subject,
+                'predicate' => $predicate,
+                'object' => $object,
+                'l_language' => is_null($lang) ? '' : $lang,
+                'author' => is_null($author) ? '' : $author,
+                'epoch' => $date,
+            ]
+        );
     }
-    
 
-}
\ No newline at end of file
+    /**
+     * Generates a unique, not auto-increment based, primary key.
+     *
+     * @return string
+     */
+    public function getUniquePrimaryKey()
```

Comment:
this stuff is in a lot of places, maybe consider a trait, of if you can, use DI for a tiny generator (will ease your life once you'll start to test)

Reply:
Right, moved to oat\generis\Helper\UuidPrimaryKeyTrait